File: controller/e_bouncing_balls.cpp
//
// Created by Nico Wittig on 2018-12-18.
//

#include "e_bouncing_balls.h"

e_bouncing_balls::e_bouncing_balls(light_element *le, uint8_t ball_count, CHSV* colors) : effect(le, EFFECT_NAME) {
    this->ball_count = ball_count;
    this->colors = (CHSV *) malloc(sizeof(CHSV) * ball_count);
    for (int ball = 0; ball < ball_count; ball++) this->colors[ball] = colors[ball];
}

e_bouncing_balls::e_bouncing_balls(light_element *le, uint8_t ball_count, CRGB *colors) : effect(le, EFFECT_NAME) {
    this->ball_count = ball_count;
    this->colors = (CHSV *) malloc(sizeof(CHSV) * ball_count);
    for (uint8_t ball = 0; ball < ball_count; ball++) this->colors[ball] = rgb2hsv_approximate(colors[ball]);
}

e_bouncing_balls::~e_bouncing_balls() {
    free(colors);

    free(Height);
    free(ImpactVelocity);
    free(TimeSinceLastBounce);
    free(Position);
    free(ClockTimeSinceLastBounce);
    free(Dampening);
}

bool e_bouncing_balls::init() {

    ImpactVelocityStart = sqrt( -2 * Gravity * StartHeight );

    Height = (float *) malloc(sizeof(float) * ball_count);
    ImpactVelocity = (float *) malloc(sizeof(float) * ball_count);
    TimeSinceLastBounce = (float *) malloc(sizeof(float) * ball_count);
    Position = (uint16_t *) malloc(sizeof(uint16_t) * ball_count);
    ClockTimeSinceLastBounce = (uint32_t *) malloc(sizeof(uint32_t) * ball_count);
    Dampening = (float *) malloc(sizeof(float) * ball_count);

    for (int ball = 0 ; ball < ball_count ; ball++) {
        ClockTimeSinceLastBounce[ball] = millis();
        Height[ball] = StartHeight;
        Position[ball] = 0;
        ImpactVelocity[ball] = ImpactVelocityStart;
        TimeSinceLastBounce[ball] = 0;
        Dampening[ball] = 0.90 - float(ball)/pow(ball_count,2);
    }

    set_all(CRGB::Black);

    return true;
}

bool e_bouncing_balls::run() {

    set_all(CRGB::Black);

    for (int ball = 0; ball < ball_count; ball++) {
        TimeSinceLastBounce[ball] = millis() - ClockTimeSinceLastBounce[ball];
        Height[ball] = 0.5 * Gravity * pow(TimeSinceLastBounce[ball] / 1000, 2.0) +
                ImpactVelocity[ball] * TimeSinceLastBounce[ball] / 1000;

        if (Height[ball] < 0) {
            Height[ball] = 0;
            ImpactVelocity[ball] = Dampening[ball] * ImpactVelocity[ball];
            ClockTimeSinceLastBounce[ball] = millis();

            if (ImpactVelocity[ball] < 0.01) {
                ImpactVelocity[ball] = ImpactVelocityStart;
            }
        }
        Position[ball] = round(Height[ball] * (le->get_num_leds() - 1) / StartHeight);

        set_pixel(Position[ball], colors[ball]);
    }

    return false;
}

File: server/e_bouncing_balls.py
import math
from time import time

from color import Color
from effect import Effect


class E_Bouncing_Balls(Effect):
    def __init__(self, light_element, ball_colors):
        super().__init__(light_element, delay=0, name="Bouncing Balls")

        self.__ball_colors = ball_colors
        self.__ball_count = len(ball_colors)

        self.__gravity = -9.81
        self.__start_height = 1

        self.__impact_velocity_start = math.sqrt(-2 * self.__gravity * self.__start_height)

        self.__height = []
        self.__impact_velocity = []
        self.__time_since_last_bounce = []
        self.__position = []
        self.__clock_time_since_last_bounce = []
        self.__damping = []

    def __get_millis(self):
        return round(time() * 1000)

    def init(self):
        for i in range(self.__ball_count):
            self.__height.append(self.__start_height)
            self.__impact_velocity.append(0.0)
            self.__time_since_last_bounce.append(0.0)
            self.__position.append(0)
            self.__clock_time_since_last_bounce.append(self.__get_millis())
            self.__damping.append(0.9 - float(i)/math.pow(self.__ball_count, 2))

        self.set_color(Color((0, 255, 0)))

        return True

    def run(self):
        if self.ready_to_run():
            self.set_color(Color((0, 255, 0)))

            for i in range(self.__ball_count):
                self.__time_since_last_bounce[i] = self.__get_millis() - self.__clock_time_since_last_bounce[i]
                self.__height[i] = 0.5 * self.__gravity * math.pow(self.__time_since_last_bounce[i] / 1000, 2) \
                                 + self.__impact_velocity[i] * self.__time_since_last_bounce[i] / 1000

                if self.__height[i] < 0:
                    self.__height[i] = 0
                    self.__impact_velocity[i] = self.__damping[i] * self.__impact_velocity[i]
                    self.__clock_time_since_last_bounce[i] = self.__get_millis()

                    if self.__impact_velocity[i] < 0.01:
                        self.__impact_velocity[i] = self.__impact_velocity_start

                self.__position[i] = round(self.__height[i] * (self._Effect__light_element.num_pixels - 1) / self.__start_height)
                p = self.__position[i]
                b = self.__ball_colors[i]
                self.set_pixel(self.__position[i], self.__ball_colors[i])

        return False
